--- old/src/java.security.jgss/share/classes/sun/security/jgss/krb5/Krb5NameElement.java	2019-10-21 14:15:13.056493848 -0500
+++ new/src/java.security.jgss/share/classes/sun/security/jgss/krb5/Krb5NameElement.java	2019-10-21 14:15:13.060493798 -0500
@@ -317,6 +317,32 @@
     }
 
     /**
+     * Returns a local username (platform-specific) corresponding to the
+     * principal name, and may return null when no username is known for the
+     * principal name.
+     *
+     * @return username corresponding to this principal name, if any
+     */
+    public String getLocalName() throws GSSException {
+        throw new GSSException(GSSException.UNAVAILABLE, -1,
+                "Mapping Kerberos principal names to usernames is not " +
+                "currently supported");
+    }
+
+    /**
+     * Returns a local username (platform-specific) corresponding to the
+     * principal name, and may return null when no username is known for the
+     * principal name.
+     *
+     * @return username corresponding to this principal name, if any
+     */
+    public String getLocalName(Oid mech) throws GSSException {
+        throw new GSSException(GSSException.UNAVAILABLE, -1,
+                "Mapping Kerberos principal names to usernames is not " +
+                "currently supported");
+    }
+
+    /**
      * Returns the name type oid.
      */
     public Oid getGSSNameType() {
@@ -341,6 +367,10 @@
         return (gssNameType.equals(GSSName.NT_ANONYMOUS));
     }
 
+    public boolean isDefaultCredentialName() {
+        return false;
+    }
+
     public Provider getProvider() {
         return Krb5MechFactory.PROVIDER;
     }
