--- old/src/java.security.jgss/share/classes/sun/security/jgss/krb5/Krb5AcceptCredential.java	2019-10-21 14:15:12.712498173 -0500
+++ new/src/java.security.jgss/share/classes/sun/security/jgss/krb5/Krb5AcceptCredential.java	2019-10-21 14:15:12.716498122 -0500
@@ -47,6 +47,7 @@
 
     private final Krb5NameElement name;
     private final ServiceCreds screds;
+    private boolean isDefCred = false;
 
     private Krb5AcceptCredential(Krb5NameElement name, ServiceCreds creds) {
         /*
@@ -57,6 +58,8 @@
 
         this.name = name;
         this.screds = creds;
+        if (name == null)
+            isDefCred = true;
     }
 
     static Krb5AcceptCredential getInstance(final GSSCaller caller, Krb5NameElement name)
@@ -148,6 +151,15 @@
         return Krb5MechFactory.GSS_KRB5_MECH_OID;
     }
 
+    /**
+     * Returns true if the credential is a default credential.
+     *
+     * @return true if the credential is a default credential, else false.
+     */
+    public boolean isDefaultCredential() {
+        return isDefCred;
+    }
+
     public final java.security.Provider getProvider() {
         return Krb5MechFactory.PROVIDER;
     }
