--- old/src/java.security.jgss/share/classes/sun/security/jgss/krb5/Krb5InitCredential.java	2019-10-21 14:15:12.820496815 -0500
+++ new/src/java.security.jgss/share/classes/sun/security/jgss/krb5/Krb5InitCredential.java	2019-10-21 14:15:12.824496765 -0500
@@ -56,6 +56,7 @@
     private Krb5NameElement name;
     private Credentials krb5Credentials;
     public KerberosTicket proxyTicket;
+    private boolean isDefCred;
 
     private Krb5InitCredential(Krb5NameElement name,
                                byte[] asn1Encoding,
@@ -88,6 +89,8 @@
         KerberosSecrets.getJavaxSecurityAuthKerberosAccess()
                 .kerberosTicketSetServerAlias(this, serverAlias);
         this.name = name;
+        if (name == null)
+            isDefCred = true;
 
         try {
             // Cache this for later use by the sun.security.krb5 package.
@@ -149,7 +152,8 @@
                 .kerberosTicketSetServerAlias(this, serverAlias);
         this.name = name;
         // A delegated cred does not have all fields set. So do not try to
-        // creat new Credentials out of the delegatedCred.
+        // creat new Credentials out of the delegatedCred.  Also, a delegated
+        // credential is not a default credential.
         this.krb5Credentials = delegatedCred;
     }
 
@@ -311,6 +315,15 @@
         return Krb5MechFactory.GSS_KRB5_MECH_OID;
     }
 
+    /**
+     * Returns true if the credential is a default credential.
+     *
+     * @return true if the credential is a default credential, else false.
+     */
+    public boolean isDefaultCredential() {
+        return isDefCred;
+    }
+
     public final java.security.Provider getProvider() {
         return Krb5MechFactory.PROVIDER;
     }
