--- old/src/java.security.jgss/share/classes/sun/security/jgss/krb5/Krb5ProxyCredential.java	2019-10-21 14:15:13.168492440 -0500
+++ new/src/java.security.jgss/share/classes/sun/security/jgss/krb5/Krb5ProxyCredential.java	2019-10-21 14:15:13.172492390 -0500
@@ -92,6 +92,11 @@
     }
 
     @Override
+    public final boolean isDefaultCredential() {
+        return false;
+    }
+
+    @Override
     public final Oid getMechanism() {
         return Krb5MechFactory.GSS_KRB5_MECH_OID;
     }
