--- old/src/java.security.jgss/share/native/libj2gss/gssapi.h	2019-10-21 14:15:15.444463827 -0500
+++ new/src/java.security.jgss/share/native/libj2gss/gssapi.h	2019-10-21 14:15:15.452463727 -0500
@@ -119,6 +119,19 @@
 typedef OM_uint32       gss_qop_t;
 typedef int             gss_cred_usage_t;
 
+/* Credential store extensions */
+typedef struct gss_key_value_element_struct {
+    const char *key;
+    const char *value;
+} gss_key_value_element_desc;
+
+typedef struct gss_key_value_set_struct {
+    OM_uint32 count; /* should be size_t, but for MIT compat */
+    gss_key_value_element_desc *elements;
+} gss_key_value_set_desc, *gss_key_value_set_t;
+
+typedef const gss_key_value_set_desc *gss_const_key_value_set_t;
+
 /*
  * Flag bits for context-level services.
  */
@@ -699,6 +712,32 @@
 #    pragma pack(pop)
 #endif
 
+/* Common extension (NOT in RFC2744) */
+OM_uint32
+gss_add_cred_with_password(
+        OM_uint32 *,            /* minor_status */
+        gss_const_cred_id_t,    /* input_cred_handle */
+        gss_const_name_t,       /* desired_name */
+        gss_const_OID,          /* desired_mech */
+        gss_const_buffer_t,     /* password */
+        gss_cred_usage_t,       /* cred_usage */
+        OM_uint32,              /* initiator_time_req */
+        OM_uint32,              /* acceptor_time_req */
+        gss_cred_id_t *,        /* output_cred_handle */
+        gss_OID_set *,          /* actual_mechs */
+        OM_uint32 *,            /* initiator_time_rec */
+        OM_uint32 *             /* acceptor_time_rec */
+);
+
+/* Common extension for aname2lname (NOT in RFC2744) */
+OM_uint32
+gss_localname(
+        OM_uint32 *,            /* minor_status */
+        gss_const_name_t,       /* name */
+        gss_OID,                /* mech_type */
+        gss_buffer_t            /* localname */
+);
+
 #ifdef __cplusplus
 }
 #endif
