--- old/src/java.security.jgss/share/native/libj2gss/NativeUtil.h	2019-10-21 14:15:15.332465235 -0500
+++ new/src/java.security.jgss/share/native/libj2gss/NativeUtil.h	2019-10-21 14:15:15.336465185 -0500
@@ -36,18 +36,21 @@
 #endif
   extern jint getJavaTime(OM_uint32);
   extern OM_uint32 getGSSTime(jint);
-  extern void checkStatus(JNIEnv *, jobject, OM_uint32, OM_uint32, char*);
+  extern void checkStatus(JNIEnv *, jobject, OM_uint32, OM_uint32, const char *);
   extern jint checkTime(OM_uint32);
   extern void throwOutOfMemoryError(JNIEnv *, const char*);
-  extern void initGSSBuffer(JNIEnv *, jbyteArray, gss_buffer_t);
-  extern void resetGSSBuffer(gss_buffer_t);
+  extern void initGSSBuffer(JNIEnv *, jbyteArray, gss_buffer_t, jboolean);
+  extern void resetGSSBuffer(JNIEnv *, jbyteArray, gss_buffer_t);
+  extern void initGSSBufferString(JNIEnv *, jstring, gss_buffer_t);
+  extern void resetGSSBufferString(JNIEnv *, jstring, gss_buffer_t);
+  extern void initGSSCredStore(JNIEnv *, jarray, gss_key_value_set_desc *);
+  extern void resetGSSCredStore(JNIEnv *, jarray, gss_key_value_set_desc *);
 
   extern gss_OID newGSSOID(JNIEnv *, jobject);
   extern void deleteGSSOID(gss_OID);
-  extern gss_OID_set newGSSOIDSet(gss_OID);
-  extern void deleteGSSOIDSet(gss_OID_set);
+  extern gss_OID_set makeGSSOIDSet(gss_OID_set, gss_OID);
 
-  extern jbyteArray getJavaBuffer(JNIEnv *, gss_buffer_t);
+  extern jbyteArray getJavaBuffer(JNIEnv *, gss_buffer_t, jboolean);
   extern jstring getJavaString(JNIEnv *, gss_buffer_t);
   extern jobject getJavaOID(JNIEnv *, gss_OID);
   extern jobjectArray getJavaOIDArray(JNIEnv *, gss_OID_set);
