--- old/src/jdk.jdwp.agent/unix/native/libjdwp/linker_md.c	2019-10-21 14:15:15.684460810 -0500
+++ new/src/jdk.jdwp.agent/unix/native/libjdwp/linker_md.c	2019-10-21 14:15:15.688460760 -0500
@@ -66,6 +66,28 @@
     free(paths_copy);
 }
 
+int
+dbgsysGetLastErrorString(char *buf, int len)
+{
+    const char *s = dlerror();
+    size_t n;
+    size_t l = (size_t)len;
+
+    if (len <= 0)
+        return 0;
+
+    *buf = '\0';
+    if (s == NULL)
+        return 0;
+
+    n = strlen(s);
+    if (n >= l)
+        n = l - 1;
+    strncpy(buf, s, n);
+    buf[n] = '\0'; /* not actually needed */
+    return n;
+}
+
 /*
  * create a string for the JNI native function name by adding the
  * appropriate decorations.
